### PR TITLE
Refactoring Exemptions

### PR DESCRIPTION
**Pull request by lucasborin**
Solves #159

### CODE REVIEW

File: src/foundation/y_object_creation_date.clas.abap [hunk 1]
```diff
@@ -87,13 +87,10 @@ CLASS Y_OBJECT_CREATION_DATE IMPLEMENTATION.
 
 
   METHOD get_created_on_from_buffer.
-    SELECT SINGLE created_on FROM ytab_exemptions INTO @creation_date
-      WHERE object                 = @object_type AND
-            obj_name               = @object_name AND
-            is_created_on_buffered = @abap_true.
-    IF sy-subrc = 0 AND creation_date IS NOT INITIAL AND creation_date <> '000000'.
```
Comment (lucasborin) [inline src/foundation/y_object_creation_date.clas.abap:94]: the creation date must be part of the xsdbool vlaidation

File: src/foundation/y_exemption_handler.clas.abap [hunk 2]
```diff
@@ -54,18 +51,8 @@ CLASS Y_EXEMPTION_HANDLER IMPLEMENTATION.
     exemption-as4date_co = sy-datum.
     exemption-is_exemption_buffered = abap_true.
 
-    INSERT ytab_exemptions FROM exemption.
-    IF sy-subrc = 0.
-      is_stored = abap_true.
-    ELSE.
-      UPDATE ytab_exemptions SET is_exempted = @is_exempted,
-                                 as4date_co = @sy-datum,
-                                 is_exemption_buffered = @abap_true
```
Comment (lucasborin) [inline src/foundation/y_exemption_handler.clas.abap:63]: Not implemented in the new version

**Comment by lucasborin** [comment]
Open points:
1. Should we remove the ytab_exemptions? If so, declare the type in the buffer class. 
2. Having this change, every execution will have their own buffer (even using multi threads). Is it better than having a buffer in logical DB?
3. Convert the exporting to returning parameters.

**Comment by larshp** [comment]
Hi,
`Y_BUFFER` is a very generic name IMHO, which risks conflicting with customer object names, suggest prefixing all objects to attempt not causing clashes, eg Y_PAL_BUFFER / Y_CPA_BUFFER

**Comment by lucasborin** [comment]
Good point @larshp!

I'll change it to Y_CODE_PAL_BUFFER, and I will check with the team if we should use a prefix like your example (y_pal, y_cpa, ...).

File: src/foundation/y_exemption_dispatcher.clas.abap [hunk 3]
```diff
@@ -35,21 +33,8 @@ CLASS y_exemption_dispatcher IMPLEMENTATION.
...
-    IF sy-subrc IS NOT INITIAL
-    OR is_dataset_outdated( exemption-as4date ) = abap_true.
-      RAISE EXCEPTION TYPE ycx_entry_not_found.
-      RETURN.
-    ENDIF.
-
-    result = exemption-is_exempted.
-
+    result = y_code_pal_buffer=>get( object_type = object_type
+                            object_name = CONV #( object_name ) )-is_exempted.
```
Comment (lucasborin) [inline src/foundation/y_exemption_dispatcher.clas.abap]: missing alignment

File: src/foundation/y_exemption_dispatcher.clas.abap [hunk 4]
```diff
@@ -115,15 +84,11 @@ CLASS y_exemption_dispatcher IMPLEMENTATION.
...
 
-    DATA(exemption) = VALUE ytab_exemptions( object      = object_type
-                                             obj_name    = object_name
-                                             is_exempted = result
-                                             as4date     = sy-datum ).
-
-    store_exemption_in_database( exemption ).
+    y_code_pal_buffer=>modify( VALUE #( object_type = object_type
+                               object_name = object_name
+                               is_exempted = result ) ).
```
Comment (lucasborin) [inline src/foundation/y_exemption_dispatcher.clas.abap]: missing alignment

**Pull request merged**